inginious/frontend/pages/course_admin/api_auth.py

```diff
@@ -0,0 +1,75 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of INGInious. See the LICENSE and the COPYRIGHTS files for
+# more information about the licensing of this file.
+
+from flask import request, Response
+import jwt
+from datetime import datetime
+import json
+from bson.json_util import dumps
+
+from inginious.frontend.pages.course_admin.utils import INGIniousAdminPage
+
+
+class CourseAPIAuthPage(INGIniousAdminPage):
+
+    def GET(self, courseid):
+        """ GET request """
+        auth_header = request.headers.get('Authorization')
+        msg, code = self.decode_token(auth_header)
+        if code == 200:
+            # DB request parameters
+            params = {"courseid": msg}
+            if "taskid" in request.args:
+                params["taskid"] = request.args.get("taskid")
+            if "username" in request.args:
+                params["username"] = request.args.get("username")
+
+            if request.args.get("type") == "tasks":
+                msg = dumps(self.database.user_tasks.find(params))
+            elif request.args.get("type") == "submissions":
+                msg = dumps(self.database.submissions.find(params))
+            else:
+                msg = json.dumps({"message": "No data type specified"})
+        else:
+            # error message
+            msg = json.dumps({"message": msg})
+
+        return self.page(msg, code)
+
+    def decode_token(self, auth_header):
+        """ Returns decoded courseid or error, and http code """
+        key = str(self.app.jwt_key)
+
+        try:
+            if auth_header[:7] != "Bearer ":
```

Comment:
Do not implement HTTP specs by yourself. Rely on external librairies that will handle this seamlessly while being updated (well, that remains a hope in practice) in the spec changes. Take a look at Flask-HTTPAuth for instance.

There may even be librairies that directly handle token authentication.